osu.Game/OsuGameBase.cs

```diff
@@ -173,6 +163,42 @@ private void load()
             FileStore.Cleanup();
         }
 
+        private void runMigrations()
+        {
+            try
+            {
+                using (var context = contextFactory.GetContext())
+                    context.Migrate();
+            }
+            catch (MigrationFailedException e)
+            {
+                Logger.Log(e.ToString(), LoggingTarget.Database, LogLevel.Error);
+                Logger.Log("Migration failed! We'll be starting with a fresh database.", LoggingTarget.Database, LogLevel.Error);
+
+                // if we failed, let's delete the database and start fresh.
+                // todo: we probably want a better (non-destructive) migrations/recovery process at a later point than this.
+                int retries = 20;
+                while (retries-- > 0)
+                {
+                    try
+                    {
+                        using (var context = contextFactory.GetContext())
+                        {
+                            context.Database.EnsureDeleted();
+                            Logger.Log("Database purged successfully.", LoggingTarget.Database, LogLevel.Important);
+                            break;
+                        }
+                    }
+                    catch
+                    {
+                    }
+                }

```

Comment:
I mean first of all the point of using `using ()` is to ensure that the objects and their underlying resources are deterministically closed. If EnsureDeleted is failing while all the resources from the previous context are supposedly closed, I want to know why.
Please provide the exception this is failing with.

Secondly if EnsureDeleted fails for these 20 attempts we're still left with an unusable database. What do?

Should we start using retries any time file access is involved?

Reply:
fails with file-in-use. unsure why, possibly sqlite holding an open file handle slightly longer than it needs? there's many layers interacting here which can lead to these kinds of issues (EF -> EF-Database -> sqlite native)

if it fails, it will attempt again the next time the user runs the app. keep in mind this isn't to do with just sqlite-net migration. this recoveres from *any* database bad states, including EF migrations, so "removing migration" is not an option.